cmd/server/run.go

```diff
@@ -37,84 +39,123 @@ func Run(ctx context.Context, conf config.Config) error {
 type Server struct {
 	*server.Server
 	*scheduler.Scheduler
-	DB server.Database
 }
 
 // NewServer returns a new Funnel server + scheduler based on the given config.
 func NewServer(conf config.Config, log *logger.Logger) (*Server, error) {
 	log.Debug("NewServer", "config", conf)
 
-	var backend compute.Backend
-	var db server.Database
-	var sdb scheduler.Database
+	ctx := context.Background()
+	var reader tes.ReadOnlyServer
+	var nodes schedProto.SchedulerServiceServer
 	var sched *scheduler.Scheduler
-	var err error
+	var queue scheduler.TaskQueue
+	writers := events.MultiWriter{}
 
 	switch strings.ToLower(conf.Server.Database) {
 	case "boltdb":
-		db, err = boltdb.NewBoltDB(conf)
+		b, err := boltdb.NewBoltDB(conf)
+		if err != nil {
+			return nil, dberr(err)
+		}
+		reader = b
+		nodes = b
+		queue = b
+		writers.Append(b)
+
 	case "dynamodb":
-		db, err = dynamodb.NewDynamoDB(conf.Server.Databases.DynamoDB)
+		d, err := dynamodb.NewDynamoDB(conf.Server.Databases.DynamoDB)
+		if err != nil {
+			return nil, dberr(err)
+		}
+		reader = d
+		writers.Append(d)
+
 	case "elastic":
-		db, err = elastic.NewTES(conf.Server.Databases.Elastic)
-	case "mongodb":
-		db, err = mongodb.NewMongoDB(conf.Server.Databases.MongoDB)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("error occurred while connecting to or creating the database: %v", err)
-	}
+		e, err := elastic.NewElastic(conf.Server.Databases.Elastic)
+		if err != nil {
+			return nil, dberr(err)
+		}
+		if err := e.Init(ctx); err != nil {
```

Comment:
why wasn't init absorbed into the constructor method for elastic like you did for the other databases?

Reply:
I guess the other option is to pass context to NewElastic. I didn't want to hunt down other NewElastic calls at the time I wrote this, so this was the quick solution